### ISSUE TITLE
Server-Side Validation

### ISSUE THREAD

**Issue by stunaz** [google/material-design-lite#1463]
Hi,
I am doing a server side validation of my form. But when I am trying to display the error on the my page by adding the class `is-invalid` while building the page (server side), this class just disappear when the page is loaded.

How can force my field to display the error? 
Do I really have to re-add the class once the page is ready (with javascript)?

**Comment by Garbee** [comment]
Are you using in-browser validation as well? Doing that would automatically provide the invalid state on load.

**Comment by stunaz** [comment]
No, I am not using the in-browser (added `novalidate` in the form tag)
here is a codepen that I created http://codepen.io/stunaz/pen/ZGgaaY?editors=100
I was expecting the field to be red.

**Comment by Garbee** [comment]
The component is checking the HTML5 validity state on initialization. This is what causes the invalid state to get removed. However, by not using the built-in validation we have no way of verifying state if it is correct once the user does input proper values. So even if things are updated to allow an initially invalid state, as soon as something changes it is then "valid" again, even if continually invalid.

Why are you not using browser validation? I'm not seeing much of a reason for us to change the library for this use-case.

**Comment by surma** [comment]
Our init process uses the browser-native validation to see if `is-invalid` should be set or not (
[`init()`](https://github.com/google/material-design-lite/blob/master/src/textfield/textfield.js#L218) :arrow_right: [_updateClasses()](https://github.com/google/material-design-lite/blob/master/src/textfield/textfield.js#L105) :arrow_right: [checkValidity()](https://github.com/google/material-design-lite/blob/master/src/textfield/textfield.js#L131). So the class will currently be removed on page load, provided the input is consider valid by the browser-native validation.

@sgomes @addyosmani Is this something we should handle upstream? Namely not removing the `is-invalid` class during `init()`.

@stunaz As a workaround, you can listen to the `mdl-componentupgraded` event and add the `is-invalid` class then.

**Comment by Garbee** [comment]
FTR handling not removing the class during init is trivial. I am simply worried about making the change for a niche use-case that is also discouraged. Developers should offer validation early and often.

According to the [error spec](https://www.google.com/design/spec/patterns/errors.html#errors-user-input-errors):

> Help users fix input errors as soon as they are detected. Disable the submission of a form if errors are detected, and if detected only after form submission, clearly explain the error and how to fix it.

By not doing validation in-browser, it goes against what MD recommends for handling errors. That is why I'm unsure we should introduce this change.

**Comment by surma** [comment]
@Garbee This is not about NOT doing validation in-browser, but *also* supporting server-side validation, which is orthogonal to our “static content only” motto. If you want to mark a field as invalid, you should be able to that, shouldn’t you? Not all validation is viable to be done client-side, imo. Also, I don’t really see this as a niche use-case.

**Comment by stunaz** [comment]
@surma thanks for suggestion, I will try it. 
A typical use-case would be on the registration form, to validate that the email/username is not already used in the database. But mostly all the business rules of an application can be validated only server-side.

**Comment by g3rd** [comment]
@Garbee I wouldn't consider this a niche use-case. I'm using Django to build out MDL HTML, almost every project I work on has additional validation that gets done server-side. Example validating that an EIN number is valid and a 501c3. An easier way to flag to the JavaScript validation that an existing error is present would be great.

**Comment by lastid** [comment]
Validation on the server side is a niche use case?!
Not everything can be validated just by using the pattern property of the input element.

In my niche use-case on the client side, I use a javascript library (react) to render the dom, so I need to call componentHandler.upgrade... The 'is-invalid' class is stripped out after this call. I got a work around, but it would be cool to fix this behavior.

I just wanted to +1 this issue, but you guys dont like that as stated in another issue. But if I and other people dont +1, this might look like a niche use case :)

**Comment by stunaz** [comment]
Or would be great to provide a way to do remove validation :+1:

**Comment by Garbee** [comment]
It is going to be looked at. Let's not blow my poor wording out of proportion and make this into something it shouldn't be. I was simply trying to say relying *only* on server-side validation is a niche case. However, the point of usernames and email uniqueness is a valid issue I wasn't considering when I wrote my initial response (even though, this can be handled client-side with custom validity.)

As far as removing validation entirely, please file a new issue. That deserves its own discussion of merit and MD alignment and I'd rather we not digress issues simply due some minor topical relationships.

**Comment by stunaz** [comment]
My bad, i meant "remote" validation

**Comment by Garbee** [comment]
Ah, yea remote/server-side, same thing.

Ideally to solve this case we just check for "is-invalid" already present and replace it after the `updateClasses_` call. This at least covers the scope of this issue just fine it appears to me from some rudimentary testing just now.

**Comment by juniebugs** [comment]
I know this old, but here's a quick hack I used to get ASP.NET/MVC jQuery validation to play nice you MDL validation.
`//MDL, jQuery Unobtrusive Validation helpers
$(document).ready(function () {
    initMdlJqueryValidation();
    showInvalidMdlTextFields();
});

var initMdlJqueryValidation = function () {
    var oldHighlight = jQuery.validator.defaults.highlight;
    var oldUnhighlight = jQuery.validator.defaults.unhighlight;

    jQuery.validator.setDefaults({
        highlight: function (element, errorClass, validClass) {

            //mdl-textfield parent div
            $(element).parent().addClass('is-invalid');

            //mdl-textfield__error span
            $(element).siblings('.mdl-textfield__error').css({ "visibility": "visible" });

            oldHighlight(element, errorClass, validClass);
        },
        unhighlight: function (element, errorClass, validClass) {

            //mdl-textfield parent div
            $(element).parent().removeClass('is-invalid');

            //mdl-textfield__error span
            $(element).siblings('.mdl-textfield__error').css({ "visibility": "hidden" });

            oldUnhighlight(element, errorClass, validClass);
        }
    });
}

var showInvalidMdlTextFields = function () {
    $('form').find('.input-validation-error.mdl-textfield__input').each(function () {

        //mdl-textfield parent div
        $(this).parent().addClass('is-invalid');

        //mdl-textfield__error span
        $(this).siblings('.mdl-textfield__error').css({ "visibility": "visible" });
    })
};`

**Comment by juniebugs** [comment]
`//MDL, jQuery Unobtrusive Validation helpers
$(document).ready(function () {
    initMdlJqueryValidation();
    showInvalidMdlTextFields();
});

var initMdlJqueryValidation = function () {
    var oldHighlight = jQuery.validator.defaults.highlight;
    var oldUnhighlight = jQuery.validator.defaults.unhighlight;

    jQuery.validator.setDefaults({
        highlight: function (element, errorClass, validClass) {

            //mdl-textfield parent div
            $(element).parent().addClass('is-invalid');

            //mdl-textfield__error span
            $(element).siblings('.mdl-textfield__error').css({ "visibility": "visible" });

            oldHighlight(element, errorClass, validClass);
        },
        unhighlight: function (element, errorClass, validClass) {

            //mdl-textfield parent div
            $(element).parent().removeClass('is-invalid');

            //mdl-textfield__error span
            $(element).siblings('.mdl-textfield__error').css({ "visibility": "hidden" });

            oldUnhighlight(element, errorClass, validClass);
        }
    });
};

var showInvalidMdlTextFields = function () {
    $('form').find('.input-validation-error.mdl-textfield__input').each(function () {

        //mdl-textfield parent div
        $(this).parent().addClass('is-invalid');

        //mdl-textfield__error span
        $(this).siblings('.mdl-textfield__error').css({ "visibility": "visible" });
    });
};`

**Comment by fenton-nick** [comment]
I know this is old, but here's a quick hack I used to get ASP.NET/MVC jQuery validation to play nice you MDL validation.

![jqueryvalidationmdl](https://user-images.githubusercontent.com/4750903/30451353-6284e1b2-9961-11e7-8116-d9c524cfc712.PNG)

**Comment by cluetjen** [comment]
Thanks @juniebugs, that worked for me. Found this issue in google, so perhaps others will too - here's a "copy paste enabled" version of the code:

```
$(document).ready(function() {
    initMdlJqueryValidation();
});

var initMdlJqueryValidation = function() {
    var oldHighlight = jQuery.validator.defaults.highlight;
    var oldUnhighlight = jQuery.validator.defaults.unhighlight;

    jQuery.validator.setDefaults({
        highlight: function(element, errorClass, validClass) {
            $(element).parent().addClass('is-invalid');
            oldHighlight(element, errorClass, validClass);
        },
        unhighlight: function(element, errorClass, validClass) {
            $(element).parent().removeClass('is-invalid');
            oldUnhighlight(element, errorClass, validClass);
        }
    });
};
```
Not that it does not set the visibility property. (It seems to work for me w/o).
